sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitions.scala

```diff
@@ -120,7 +120,7 @@ private[sql] object PruneFileSourcePartitions
 
     case op @ PhysicalOperation(projects, filters,
         v2Relation @ DataSourceV2ScanRelation(_, scan: FileScan, output))
-        if filters.nonEmpty && scan.readDataSchema.nonEmpty =>
```

Comment:
I think the above condition should be removed (see SPARK-35985).
But I might be wrong.

@cloud-fan @gengliangwang just fyi, as you wrote this [code 2 years ago](https://github.com/apache/spark/commit/94fc0e3235162afc6038019eed6ec546e3d1983e).

Reply:
I think we should check if partitionSchema is nonEmpty? This looks like a mistake.

Reply:
@cloud-fan the non-emptiness is indirectly checked 4 lines down on line 128.

Do you want me to add it here again, to prevent `getPartitionKeyFiltersAndDataFilters()` from being called?
`getPartitionKeyFiltersAndDataFilters` is able to handle `scan.readPartitionSchema` being empty.  So it would only be a performance optimization I think.

I do not mind to add it, just want to know your opinion.

Reply:
I think it's better to match the v1 code above: `if filters.nonEmpty && fsRelation.partitionSchemaOption.isDefined`

Reply:
@cloud-fan I tried what you proposed and added `&& scan.readPartitionSchema.nonEmpty`.
Problem is that this prevents any data filter from being pushed down when there is no partition filter.  As the right part of the condition at line 128 ` || (dataFilters.nonEmpty && scan.dataFilters.isEmpty)` can then never be true.

This also causes [some regression tests in tue avroSuite to fail](https://github.com/steven-aerts/spark/runs/3007716757?check_suite_focus=true). 

So I rolled back to the original proposal.

Is this ok for you?

Reply:
I see, data filters can also be pushed here.